models/balance.js

```diff
@@ -137,19 +137,51 @@ async function rateContent({ contentId, contentOwnerId, fromUserId, transactionT
   };
 }
 
-async function undo(operationId, options = {}) {
-  const balanceOperation = await findOne(operationId, options);
-
-  const invertedBalanceOperation = {
-    balanceType: balanceOperation.balance_type,
-    recipientId: balanceOperation.recipient_id,
-    amount: balanceOperation.amount * -1,
-    originatorType: 'event',
-    originatorId: options.event.id,
+async function undo({ values = {}, where }, options = {}) {
+  let globalIndex = 1;
+  const selectFields = getSelectFields();
+  const whereClause = buildWhereClause();
+
+  const query = {
+    text: `
+      INSERT INTO balance_operations
+        (balance_type, recipient_id, amount, originator_type, originator_id)
+      SELECT
+        ${selectFields.text}
+      FROM
+        balance_operations
+      ${whereClause}
+      ;`,
+    values: [...selectFields.values, ...Object.values(where)],
   };
 
-  const newBalanceOperation = await create(invertedBalanceOperation, options);
-  return newBalanceOperation;
+  await database.query(query, options);
+
+  function getSelectFields() {
+    const text = `balance_type,
+        recipient_id,
+        -amount,
+        ${values.originator_type ? `$${globalIndex++}` : 'originator_type'},
+        ${values.originator_id ? `$${globalIndex++}` : 'originator_id'}`;
+
+    const selectValues = [];
+    if (values.originator_type) {
+      selectValues.push(values.originator_type);
+    }
+    if (values.originator_id) {
+      selectValues.push(values.originator_id);
+    }
+
+    return { text, values: selectValues };
+  }
+
+  function buildWhereClause() {
+    const conditions = Object.entries(where).map(([key, value]) => {
+      return Array.isArray(value) ? `(${key} = ANY ($${globalIndex++}))` : `(${key} = $${globalIndex++})`;
+    });
+
+    return `WHERE ${conditions.join(' AND ')}`;
+  }
```

Comment:
Esse PR envolveu várias criações de "where". Eu estava fazendo outra tarefa e passei por uma situação assim novamente, mas enxerguei algo a mais.

Para a função ter uma interface mais consistente, acha que faz sentido colocar dentro do `buildWhereClause` um mapeamento do nome do atributo do objeto para o nome da coluna do banco de dados? Por exemplo:

```suggestion
  function buildWhereClause() {
    const columnMap = {
      balanceType: 'balance_type',
      contentId: 'content_id',
      originatorType: 'originator_type',
      originatorId: 'originator_id',
    }
  
    const conditions = Object.entries(where).map(([key, value]) => {
      const column = columnMap[key] ?? key;
      return Array.isArray(value) ? `${column} = ANY ($${globalIndex++})` : `${column} = $${globalIndex++}`;
    });

    return `WHERE ${conditions.join(' AND ')}`;
  }
```

Dessa forma, a função pode ser chamada sem que quem chamou precise saber o nome das tabelas no banco.

Reply:
Faz sentido!